<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>A complete LOPIMS pipeline — lopims • pRoloc</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="A complete LOPIMS pipeline — lopims" />

<meta property="og:description" content="The function processes MSe data using the synergise
function of the synapter package and combines resulting
Synapter instances into one &quot;MSnSet&quot;
and organelle marker data is added as a feature-level annotation variable." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pRoloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v01-pRoloc-tutorial.html">Using pRoloc for spatial proteomics data analysis</a>
    </li>
    <li>
      <a href="../articles/v02-pRoloc-ml.html">Machine learning techniques available in pRoloc</a>
    </li>
    <li>
      <a href="../articles/v03-pRoloc-bayesian.html">Bayesian Analysis of Spatial Proteomics data using pRoloc</a>
    </li>
    <li>
      <a href="../articles/v04-pRoloc-goannotations.html">Annotating spatial proteomics data</a>
    </li>
    <li>
      <a href="../articles/v05-pRoloc-transfer-learning.html">A transfer learning algorithm for spatial proteomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lgatto/pRoloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A complete LOPIMS pipeline</h1>
    <small class="dont-index">Source: <a href='https://github.com/lgatto/pRoloc/blob/master/R/lopims.R'><code>R/lopims.R</code></a></small>
    <div class="hidden name"><code>lopims.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The function processes MSe data using the <code><a href='https://www.rdocumentation.org/packages/synapter/topics/synergise'>synergise</a></code>
function of the <code><a href='https://www.rdocumentation.org/packages/synapter/topics/synapter-package'>synapter</a></code> package and combines resulting
<code><a href='https://www.rdocumentation.org/packages/synapter/topics/Synapter'>Synapter</a></code> instances into one <code>"MSnSet"</code>
and organelle marker data is added as a feature-level annotation variable.</p>
    
    </div>

    <pre class="usage"><span class='fu'>lopims</span>(<span class='kw'>hdmsedir</span> <span class='kw'>=</span> <span class='st'>"HDMSE"</span>, <span class='kw'>msedir</span> <span class='kw'>=</span> <span class='st'>"MSE"</span>, <span class='kw'>pep3ddir</span> <span class='kw'>=</span> <span class='st'>"pep3D"</span>,
  <span class='no'>fastafile</span>, <span class='no'>markerfile</span>, <span class='kw'>mfdr</span> <span class='kw'>=</span> <span class='fl'>0.025</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>hdmsedir</th>
      <td><p>A <code>character</code> identifying the directory
containing the HDMSe final peptide files. Default is <code>HDMSe</code>.</p></td>
    </tr>
    <tr>
      <th>msedir</th>
      <td><p>A <code>character</code> identifying the directory
containing the MSe final peptide files. Default is <code>MSe</code>.</p></td>
    </tr>
    <tr>
      <th>pep3ddir</th>
      <td><p>A <code>character</code> identifying the directory
containing the MSe pep 3D files. Default is <code>pep3D</code>.</p></td>
    </tr>
    <tr>
      <th>fastafile</th>
      <td><p>A <code>character</code> identifying the protein
fasta database. Default is to use the fasta file in the current
directory. If several such files exist, the function reports an
error.</p></td>
    </tr>
    <tr>
      <th>markerfile</th>
      <td><p>A <code>character</code> identifying the marker
file (see details for format). Default is to use a <code>csv</code> file
starting with <code>marker</code> in the current directory.
If several such files exist, the function reports an error.</p></td>
    </tr>
    <tr>
      <th>mfdr</th>
      <td><p>The master FDR value. Default is 0.025.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional paramters passed to <code><a href='https://www.rdocumentation.org/packages/synapter/topics/synergise'>synergise</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An instance of class <code>"MSnSet"</code> with protein
level quantitation and respective organelle markers.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>LOPIMS</code> pipeline is composed of 5 steps:</p>
<ol>
<li><p>The HDMSe final peptide files are used to compute false
discovery rates uppon all possible combinations of HDMSe final
peptides files and the best combination smaller or equal to
<code>mfdr</code> is chosen.  See
<code><a href='https://www.rdocumentation.org/packages/synapter/topics/estimateMasterFdr'>estimateMasterFdr</a></code> for details.  The
corresponding master run is then created as descibed in
<code><a href='https://www.rdocumentation.org/packages/synapter/topics/makeMaster'>makeMaster</a></code>. (function <code>lopims1</code>)</p></li>
<li><p>Each MSe/pep3D pair is processed using the HDMSe master file
using <code><a href='https://www.rdocumentation.org/packages/synapter/topics/synergise'>synergise</a></code>. (function <code>lopims2</code>)</p></li>
<li><p>The respective peptide-level <code>synergise</code> output objects
are converted and combined into an single
<code>"MSnSet"</code> instance. (function <code>lopims3</code>)</p></li>
<li><p>Protein-level quantitation is inferred as follows.  For each
protein, a reference sample/fraction is chosen based on the number
of missing values (<code>NA</code>). If several samples have a same
minimal number of <code>NA</code>s, ties are broken using the sum of
counts.  The peptides that do not display any missing values for
each (frac_i, frac_ref) pair are summed and the ratio is
reported (see pRoloc:::refNormMeanOfNonNAPepSum for
details). (function <code>lopims4</code>)</p></li>
<li><p>The markers defined in the <code>markerfile</code> are collated as
feature meta-data in the <code>markers</code> variable.  See
<code><a href='addMarkers.html'>addMarkers</a></code> for details. (function <code>lopims5</code>)</p></li>
</ol>
    <p>Intermediate <code>synergise</code> reports as well as resulting objects
are stored in a <code>LOPIMS_pipeline</code> directory.
For details, please refer to the <code>synapter</code> vignette and
reference papers.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Improving qualitative and quantitative performance for
    MSE-based label free proteomics N.J. Bond, P.V. Shliaha,
    K.S. Lilley and L. Gatto Journal of Proteome Research,
    2013;12(6):2340-53. PMID: 23510225.</p>
    <p>The Effects of Travelling Wave Ion Mobility Separation on Data
    Independent Acquisition in Proteomics Studies P.V. Shliaha,
    N.J. Bond, L. Gatto and K.S. Lilley Journal of Proteome
    Research, 2013;12(6):2323-39. PMID: 23514362.</p>
<p>MSnbase-an R/Bioconductor package for isobaric tagged mass
    spectrometry data visualization, processing and
    quantitation. L. Gatto and KS. Lilley. Bioinformatics. 2012
    Jan 15;28(2):288-9.  doi: 10.1093/bioinformatics/btr645. Epub
    2011 Nov 22. PubMed PMID: 22113085.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
          </ul>

    <h2>Author</h2>
    <p>Laurent Gatto</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Laurent Gatto, Lisa Breckels, Oliver Crook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

